src/js/components/componentDiagram/graph/index.js

```diff
@@ -0,0 +1,308 @@
+  animation: {
+    enable: true,
+    duration: 300,
+  },
+};
+
+export default class Graph {
+  constructor(element, options = {}) {
+    this.element = element;
+    this.options = deepmerge(DEFAULT_OPTIONS, options);
+
+    this.outputGroup = createSVGElement('g', 'output');
+    this.edgesGroup = createSVGElement('g', 'edgePaths');
+    this.clustersGroup = createSVGElement('g', 'clusters');
+    this.nodesGroup = createSVGElement('g', 'nodes');
+
+    this.outputGroup.appendChild(this.clustersGroup);
+    this.outputGroup.appendChild(this.edgesGroup);
+    this.outputGroup.appendChild(this.nodesGroup);
+
+    this.element.innerHTML = '';
+    this.element.appendChild(this.outputGroup);
+
+    this.graph = new dagre.graphlib.Graph({ compound: true }).setGraph({ rankdir: 'LR' }).setDefaultEdgeLabel(function() { return {}; });
+  }
+
+  setNode(data, parentId = null) {
+    const node = prepareNode(data);
+    // create dummy <g class="node"> with label to determine label width
+    const dummyNodeGroup = createSVGElement('g', 'node');
+    const labelGroup = new LabelGroup(node.label, true);
+    dummyNodeGroup.appendChild(labelGroup.element);
+    this.nodesGroup.appendChild(dummyNodeGroup);
+    const labelBBox = labelGroup.getBBox();
+    this.nodesGroup.removeChild(dummyNodeGroup);
+
+    node.labelWidth = labelBBox.width;
+    node.labelHeight = labelBBox.height;
+    node.width = labelBBox.width + NODE_PADDING_HORIZONTAL * 2;
+    node.height = labelBBox.height + NODE_PADDING_VERTICAL * 2;
+
+    this.graph.setNode(node.id, node);
+
+    if (parentId) {
+      this.graph.setParent(node.id, parentId);
+    }
+  }
+
+  removeNode(id) {
+    const edges = this.graph.nodeEdges(id);
+    const node = this.graph.node(id);
+
+    if (edges) {
+      edges.forEach(({ v, w }) => {
+        const edge = this.graph.edge(v, w);
+        if (edge.group) {
+          edge.group.remove();
+        }
+        this.graph.removeEdge(v, w);
+      });
+    }
+
+    node.group.remove();
+    this.graph.removeNode(id);
+  }
+
+  setEdge(start, end) {
+    if (start !== end) {
+      if (!this.graph.node(start)) {
+        this.setNode({ id: start, type: 'class' });
+      }
+
+      if (!this.graph.node(end)) {
+        this.setNode({ id: end, type: 'class' });
+      }
+
+      this.graph.setEdge(start, end);
+    }
+  }
+
+  render() {
+    dagre.layout(this.graph);
+
+    this.graph.nodes().forEach((id) => {
+      const node = this.graph.node(id);
+
+      if (node.group) {
+        node.group.move(node.x, node.y);
+        return;
+      }
+
+      if (node.type === 'cluster') {
+        const clusterGroup = new ClusterGroup(node);
+        node.group = clusterGroup;
+        this.clustersGroup.appendChild(clusterGroup.element);
+      } else {
+        const nodeGroup = new NodeGroup(node, this.options.animation);
+
+        node.group = nodeGroup;
+        node.element = nodeGroup.element;
+
+        this.nodesGroup.appendChild(nodeGroup.element);
```

Comment:
New nodes and edges don't get a chance to animate. I think it'd be most intuitive if the nodes spawned stacked on their parent (or somewhere amidst their children) and tweened to their final position.

Ideally, nothing ever instantly snaps into position or changes visibility.